### ISSUE TITLE
Cross-platform support

### ISSUE THREAD

**Issue by reflectronic** [lordmilko/ClrDebug#9]
Hey, 

We've been using this library to write our own .NET debugger and have thoroughly enjoyed using it so far.  Unfortunately, from what I can tell, the library really only works on Windows. This is primarily because of the use of `ComImport` (among other things). We'd like to use this library cross-platform and are willing to help out where possible/necessary :-)

One option here would be to take advantage of the [COM interface source generator](https://github.com/dotnet/runtime/issues/66674) that's coming with .NET 8. The infrastructure used by the source generator is cross-platform (and can also be used "[manually](https://learn.microsoft.com/en-us/dotnet/standard/native-interop/tutorial-comwrappers)" if that's desired). The nice thing about this option is that the majority of the existing code doesn't need to change at all—a find-and-replace for `ComImport` with `GeneratedComInterface` should get you 90% of the way there.

Since the generator will only be available when targeting .NET 8, it's a little unclear what the targeting situation should be like. The two interop mechanisms are not really compatible, so multi-targeting seems problematic. Perhaps you'd need to `#if` the attributes and produce separate packages?

Curious to hear what you think. Thanks.

**Comment by lordmilko** [comment]
Hi @reflectronic,

You are right that this library is mostly targeted at Windows; cross-platform isn't so much "unsupported", rather it's "untested"

ClrDebug does include a variety of definitions pertaining to .NET Core/dbgshim which would *theoretically* allow cross-platform use. From your message, it sounds like you attempted to use ClrDebug on macOS/Linux and there was some sort of issue relating to the use of the `[ComImport]` attribute? If so, what happened exactly?

Around the beginning of this year when I added all the profiling interfaces to ClrDebug I was also investigating the potential for using NativeAOT + ClrDebug for writing .NET Profilers. In order to use ClrDebug in NativeAOT however I will need to have something that generates RCWs for all the relevant interfaces in ClrDebug - and the state of the proposed source generator feature did not strike me as being anywhere near ready for primetime use. The source generator is only really needed for NativeAOT however because the subsystem in .NET that generates RCWs won't exist in your application. So as alluded to above, I *would* imagine that (barring any other issues) you may be able to use COM interfaces cross-platform already.

From a targeting perspective, from my perspective it is mandatory that .NET Standard 2.0 is supported so .NET Framework and .NET Core users can use ClrDebug with no issues, with this target never going "out of support". If need be I can add a secondary target for .NET 8 + any `#if` statements to hack things into shape as well, and then both of these targets will be available in the NuGet package.

However, it's not clear to me that the proposed .NET 8 COM enhancements are ready for use yet. Potentially we to wait for these features to actually ship in a full release of .NET before looking at how to implement this into ClrDebug

**Comment by lordmilko** [comment]
I did some testing to see whether ClrDebug works cross-platform myself.

I wrote a simple application to create an `ICorDebug` from DbgShim.

* When I attempt to call `Marshal.GetDelegateForFunctionPointer` against `CreateDebuggingInterfaceFromVersionExDelegate` I get a `MarshalDirectiveException`: `Cannot marshal 'parameter 3': Invalid managed/unmanaged type combination (Marshaling to and from COM interface pointers isn't supported).'`

* That error message is defined [here](https://github.com/dotnet/runtime/blob/e082cc63cf2f19fa41f520d65406b20f3cd3202f/src/coreclr/dlls/mscorrc/mscorrc.rc#L531) as `IDS_EE_OBJECT_TO_ITF_NOT_SUPPORTED`

* `IDS_EE_OBJECT_TO_ITF_NOT_SUPPORTED` is returned [here](https://github.com/dotnet/runtime/blob/e082cc63cf2f19fa41f520d65406b20f3cd3202f/src/coreclr/vm/mlinfo.cpp#L2093) when `FEATURE_COMINTEROP` is not defined

* `FEATURE_COMINTEROP` is only [defined](https://github.com/dotnet/runtime/blob/main/src/coreclr/clrdefinitions.cmake#L92) on Windows builds

So while the .NET Unmanaged API may use "fake COM", we can't even do that since non-Windows builds aren't including the COM infrastructure. As such it would seem we'll have to wait for the COM source generation feature to become available, which - I would hope - will allow us to bypass this issue.

**Comment by reflectronic** [comment]
Right, the built-in COM interop system does not exist at all on non-Windows platforms or on Native AOT. The newer COM interop system ([`System.Runtime.InteropServices.ComWrappers`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.comwrappers)) does work on those platforms, but it's super barebones—you need to generate RCWs and CCWs yourself (ComWrappers just handles the object lifetimes). The COM source generator uses this underneath but retains the familiar `ComImport`-style programming model.

> I can't think of any obvious reasons why multi-targeting would cause an issue here.

The risk is, admittedly, marginal, but consider the following:
- A user creates a library targeting netstandard2.0 that uses ClrDebug
  - The user sees that the interop interfaces use `ComImport`, and so they use features that only work with built-in COM interop (e.g. `Marshal` APIs) in their library
- Another user references this library from their application targeting net8.0
  - NuGet selects the net8.0 version of ClrDebug when building the application, but since the net8.0 version isn't using the built-in COM interop, the API calls made by the netstandard2.0 library fail!

This is why, in general, assets for older targets in NuGet packages need to be forward compatible with the assets for newer targets. Separating the targets into separate packages forces people to multi-target. But I understand that it's painful and niche. I suppose, if you are writing a .NET debugger, you can figure out how to follow directions.

> it's not clear to me that the proposed .NET 8 COM enhancements are ready for use yet

The COM source generator has just become usable within the past month or so. There should be enough of it available in .NET 8 Preview 5 to give it a spin. I can take a look to see what it would take...

**Comment by lordmilko** [comment]
Thanks @reflectronic,

So you're saying if `ClrDebug.nupkg` multi-targets `netstandard2.0` and `net8.0`, and we have the scenario

    ClrDebug (netstandard2.0) <- DebugHelper (netstandard2.0) <- MyDebugger (net8.0)

then `MyDebugger` will end up with the `netstandard2.0` version of ClrDebug, rather than the `net8.0` one. While this certainly seems like an issue, there aren't really any good solutions either.

Suppose ClrDebug has two packages
* `ClrDebug.NetStandard`
* `ClrDebug.Net8`

If `DebugHelper` targets `ClrDebug.NetStandard`, you're still in exactly the same boat. Even if you install `ClrDebug.Net8` in `MyDebugger`, it won't help you because all the calls `DebugHelper` makes are going to go into `ClrDebug.NetStandard` which isn't cross-platform.

I attempted to do some research on how dependency resolution works with regards to multi-targeting; I was able to find that when there are different dependency *versions*, the [direct dependency wins](https://learn.microsoft.com/en-us/nuget/concepts/dependency-resolution#direct-dependency-wins). The documentation states the *direct dependency wins* rule only applies when there is a version difference. I'll need to test whether
* it can also apply generally even if there isn't a version difference, and
* whether this rule can be used to upgrade the copy of ClrDebug that `DebugHelper` depends on from netstandard2.0 to net8.0.

While this may be the most ideal scenario, if retargeting transitive dependencies isn't possible, given you can potentially have a mess no matter what strategy you take I'll be most inclined to simply have a single multi-targeting NuGet package.

---

> The COM source generator has just become usable within the past month or so. There should be enough of it available in .NET 8 Preview 5 to give it a spin. I can take a look to see what it would take...

Are you able to provide a citation on this? I attempted to confirm this however couldn't find any results, and don't see any mention of this in the [.NET 8 Preview 5 release notes](https://devblogs.microsoft.com/dotnet/announcing-dotnet-8-preview-5/)

**Comment by reflectronic** [comment]
> then MyDebugger will end up with the netstandard2.0 version of ClrDebug, rather than the net8.0 one. 

No, it will end up with the net8.0 version. (NuGet picks the closest assets based on the application's target framework.) But DebugHelper was programmed against the netstandard2.0 version. The author of DebugHelper may have assumed that you can do `(ICorDebug)Marshal.GetObjectForIUnknown(pUnk)`, since the netstandard2.0 version of `ICorDebug` uses `ComImport`. But this code does not work if `ICorDebug` is using `GeneratedComInterface` (which it will be at runtime).

> If DebugHelper targets ClrDebug.NetStandard, you're still in exactly the same boat.

Yeah, you would be at the mercy of DebugHelper properly multitargeting itself. In fairness, the code wouldn't be broken in this case (like it would be in the case I described), it just wouldn't work cross-platform. 

As long as the code doesn't assume that it's dealing with `ComImport`, though, the interfaces themselves are still binary compatible. (And, in general, I don't think you'd ever need to assume this.) So it probably isn't worth the complexity of creating separate packages. Just don't do the wrong thing if you are making a library that depends on ClrDebug.

> Are you able to provide a citation on this?

It isn't officially documented or announced anywhere, I have just been looking at the progress. For example, they have recently started replacing ComImport usage in system libraries with the source generator (e.g. https://github.com/dotnet/runtime/pull/87351). 

I have also just tried it with ClrDebug and does initially appear to function (haven't fully tested everything). At the moment, it would require a lot of source churn because it's more strict with the marshaling annotations (e.g. `[In] int x` or `[Out] int x` or `[In] int[] x` are all disallowed because the attribue is redundant) and has a few missing features (e.g. `[MarshalAs(UnmanagedType.LPStruct)] Guid g` doesn't work yet). I might open an issue to see if some of these can be reduced to warnings/implemented.

**Comment by lordmilko** [comment]
> (ICorDebug)Marshal.GetObjectForIUnknown(pUnk)

I would expect that with the new generated COM interop system, attempting to do this will result in your custom generated COM handler being called upon. Is that not the case?

> For example, they have recently started replacing ComImport usage in system libraries with the source generator (e.g. https://github.com/dotnet/runtime/pull/87351).

Thanks! I'll take a look

> At the moment, it would require a lot of source churn because it's more strict with the marshaling annotations (e.g. [In] int x or [Out] int x or [In] int[] x are all disallowed because the attribue is redundant) and has a few missing features (e.g. [MarshalAs(UnmanagedType.LPStruct)] Guid g doesn't work yet)

All the COM interface method parameters in ClrDebug are automatically managed via my unreleased ComWrapper library (not to be confused with .NET's coincidentally named ComWrappers system). If the generated COM system doesn't like it when you use `System.Runtime.InteropServices.InAttribute`, I can potentially stub this out by defining my own `InAttribute`, etc in .NET 8 builds that would hopefully be ignored by the generated COM system. I'll check it out

**Comment by lordmilko** [comment]
Attempting to do `Marshal.GetDelegateForFunctionPointer` on a delegate containing a COM interface parameter - even if that interface type is decorated with `GeneratedComInterfaceAttribute` - still causes a `MarshalDirectiveException` to be thrown. I imagine what needs to occur here is rather than use `Marshal` to get the delegate, `ComWrrappers` needs to be used (which would also answer my question above about `Marshal.GetObjectForIUnknown`).

I don't see any method on the `ComWrappers` type that would serve as an appropriate substitute for `Marshal.GetDelegateForFunctionPointer`. The only "workaround" here would be to declare the delegate as taking an `IntPtr` rather than a specific interface, and then using `ComWrappers` to manually convert that `IntPtr` into a specific interface.

My expectation is that .NET should transparently dispatch from `Marshal` to the generated `ComWrappers` code. Are you potentially able to get `Marshal.GetDelegateForFunctionPointer` working against a delegate like

```c#
public delegate HRESULT CreateDebuggingInterfaceFromVersionExDelegate(
        [In] CorDebugInterfaceVersion iDebuggerVersion,
        [MarshalAs(UnmanagedType.LPWStr), In] string szDebuggeeVersion,
        [MarshalAs(UnmanagedType.Interface), Out] out test.ICorDebug ppCordb);
```
?

**Comment by lordmilko** [comment]
While I have started adding support for multi-targeting .NET Standard 2.0 + .NET 8.0, it appears adding cross-platform support is blocked by https://github.com/dotnet/runtime/issues/82581, with many methods generating compiler errors

```
error CS8977: Cannot use 'ref', 'in', or 'out' in the signature of a method attributed with 'UnmanagedCallersOnly'
```
As the aforementioned issue alludes, it is erroneous to suggest the code we are generating is for "unmanaged callers only". We _have_ to support emitting values as `ref` or `out` - something that is completely separate from any `System.Runtime.InteropServices` the method parameters may be decorated with.

**Comment by reflectronic** [comment]
> I would expect that with the new generated COM interop system, attempting to do this will result in your custom generated COM handler being called upon.

It wouldn't; the COM source generator doesn't hook into the built-in interop system. Though, now that I think of it, .NET has this wonderful "feature" where any interface, even if it doesn't have `ComImport`, can be used in COM interop. So I think it would probably end up working anyway by accident (though only on Windows).

> I'm able to manually control how these values are marshalled

Yeah, I think this is how you'd do this in the meantime. In the future, they might work on a way to use the source-generated interop with function pointers (see https://github.com/dotnet/runtime/issues/63590), but today you'd probably want to manually use [`System.Runtime.InteropServices.Marshalling.ComInterfaceMarshaller<T>`](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.marshalling.cominterfacemarshaller-1?view=net-8.0) as a replacement for `GetObjectFromIUnknown`.

> it appears adding cross-platform support is blocked by https://github.com/dotnet/runtime/issues/82581, with many methods generating compiler errors

This is curious, can you give an example of a method that generates an error like that? The issue suggests that the compile error, while suboptimal, is the result of an invalid marshalling configuration anyway. The `UnmanagedCallersOnly` stub method is used for the CCW side, so maybe those methods are using something that doesn't yet support "unmanaged to managed" marshalling?

**Comment by lordmilko** [comment]
Dozens of parameters in `IMetaDataImport`

```c#
[System.Runtime.InteropServices.Marshalling.GeneratedComInterfaceAttribute]
[InterfaceType(ComInterfaceType.InterfaceIsIUnknown)]
[Guid("7DAC8207-D3AE-4C75-9B67-92801A497D44")]
public partial interface IMetaDataImport
```
My interpretation of https://github.com/dotnet/runtime/issues/82581 is that the generator is not distinguishing between methods designed for RCWs vs CCWs correctly, and is erroneously applying `UnmanagedCallersOnly` to methods that are meant to be consumed by RCWs

**Comment by reflectronic** [comment]
Hm, I just converted it manually and I'm not seeing errors like that... can you send the whole 'new' file?

**Comment by lordmilko** [comment]
I have pushed my commit adding .NET 8 support

* Please re-clone or download the latest commits such that 294682f is the latest commit
* Rename the attached file back to IMetaDataImport.cs and replace the file `ClrDebug\ClrDebug\Native\Interfaces\Metadata\IMetadataImport.txt`
* Open a command prompt under `ClrDebug\ClrDebug` and execute the following command

    `dotnet build /p:MultiBuild=true`

I get a million errors. Some of the errors relate to the use of `[In]` but then there are heaps down the bottom that say

```
C:\ClrDebug\ClrDebug\Microsoft.Interop.ComInterfaceGenerator\Microsoft.Interop.ComInterfaceGenerator\ClrDebug.IMetaDataImport.cs(2170,213):
error CS8977: Cannot use 'ref', 'in', or 'out' in the signature of a method attributed with 'UnmanagedCallersOnly'. [C:\ClrDebug\ClrDebug\ClrDebug.csproj::TargetFramework=net8.0]
```


[IMetaDataImport.txt](https://github.com/lordmilko/ClrDebug/files/11765928/IMetaDataImport.txt)

**Comment by reflectronic** [comment]
Yeah, I think those errors happen because the source generator doesn't generate "clean" code if it's going to error out anyway. To get it to work, you need to remove all of the redundant marshalling attributes, and you also need to replace StringBuilder—it looks like the source-generated interop doesn't support StringBuilder (and the documentation now [recommends against using it](https://learn.microsoft.com/en-us/dotnet/standard/native-interop/best-practices#string-parameters) even for built-in interop).

**Comment by lordmilko** [comment]
I'm not convinced .NET's COM marshalling is anywhere near ready for use. I've made a copy of ClrDebug, deleted the entire `Managed` folder, all interfaces but `IMetaDataImport`, anything else referencing deleted types and have been removing all the various marshalling attributes, and replaced `StringBuilder` with `char[]` but I still get ridiculous errors like

```
Cannot use 'string' as a parameter type on a method attributed with 'UnmanagedCallersOnly'
```
```
Cannot use 'char[]' as a parameter type on a method attributed with 'UnmanagedCallersOnly'
```
Worst of all, contradictorily, after removing all the `MarshalAsAttribute` attributes to clear errors like

```
The specified 'MarshalAsAttribute' configuration for parameter 'szField' is not supported by source-generated COM. If the specified configuration is required, use `ComImport` instead
```
I instead get a heap of errors telling me to put them back!
```
Marshalling an array from unmanaged to managed requires either the 'SizeParamIndex' or 'SizeConst' fields to be set on a 'MarshalAsAttribute' or the 'ConstantElementCount' or 'CountElementName' properties to be set on a '
MarshalUsingAttribute'. The generated source will not handle marshalling of parameter 'szName'
```
If you fork ClrDebug and demonstrate the correct configuration required to get this compile I'm happy to continue looking at this, otherwise I think it is best to wait for Microsoft to properly ship and document what exactly is required to use their new system. The fact they're blacklisting the normal marshalling attributes makes no sense to me - not only does this break backwards compatibility, but if they aren't actually used by the source generated system, the attributes should just be ignored.

**Comment by lordmilko** [comment]
Hi @reflectronic,

I came across a document describing the marshalling rules for the LibraryImportGenerator released in .NET 7, which appear to be many of the same rules that the new generated COM marshalling uses, and has allowed me to bypass some of the issues I was having previously.

I have made a new issue #10 for tracking the progress of adding cross-platform/NativeAOT support.

Thanks!

**Comment by reflectronic** [comment]
Btw, I asked around, and there's a patch out that would allow `In`/`Out` to be present even if unnecessary: https://github.com/dotnet/runtime/pull/87774

So in the future there will be no need to stub out the attributes.

**Comment by lordmilko** [comment]
Hi @reflectronic,

Please be advised a pre-release version of ClrDebug 0.3.0 is now available that is cross-platform compatible!

To access pre-release versions of ClrDebug, please execute the following commands in your project

```cmd
dotnet new nugetconfig
dotnet nuget add source -n clrdebug https://ci.appveyor.com/nuget/clrdebug-l1bkrkdwc559
```
You can then install the ClrDebug 0.3.0-alpha1 build using the NuGet Package Manager in Visual Studio (be sure to tick *Include prerelease*) or by adding the following line package reference and doing `nuget restore`

```xml
<PackageReference Include="ClrDebug" Version="0.3.0-alpha1" />
```
* This pre-release version has been tested with the .NET 8 [daily build](8.0.100-preview.7.23364.32) `8.0.100-preview.7.23364.32` which contains several issues I have reported and have been fixed since .NET 8 Preview 6 was released; I believe most or all of these issues are build-time issues, so this build may work just fine using the Preview 6 runtime.

* As Microsoft will not ship a source generator for generating native versions of structs compatible with source generated COM, I have developed my own source generator for ClrDebug for the purposes of managing structs containing complex fields (`string`, `bool`, `interface`, `ByValArray`)
* Marshalling `VARIANT` parameters is not yet supported. As Microsoft's source generator does not yet handle `VARIANT`, ClrDebug has to define its own custom `VARIANT` marshaller without relying on native methods like `VariantClear`. There is a little bit of complexity to this so I've put a pin in this for now
* Currently, only types in the ClrDebug namespace are cross-platform/NativeAOT compatible. Types in the `ClrDebug.DbgEng` do not use source generated COM in .NET 8.0 builds yet. DbgEng obviously isn't cross-platform compatible anyway, but on Windows you could still use ClrDebug to write a NativeAOT compiled DbgEng extension
* I am still investigating to what extent an `rd.xml` file may be required for interacting with certain delegates in NativeAOT builds
* I have successfully tested ClrDebug 0.3.0-alpha1 on Linux, and was able to retrieve an `ICorDebug` from `DbgShim` and have my `ICorDebugManagedCallback` list all modules as they are loaded
* There have been some changes to the API surface of `DbgShim`, and more may potentially be made (`protected virtual T GetDelegate<T>(string procName)` may potentially go away). The most important thing to note is that you can get a `CorDebug` wrapper straight from the delegate passed to `RegisterForRuntimeStartup` now. More info about this can be found [here](https://github.com/lordmilko/ClrDebug/blob/master/Samples/NetCore/Program.cs#L155)

Please let me know if you find any issues
